storage_client.go

```diff
@@ -13,9 +23,137 @@ type storageUtil interface {
+}
+
+func (rsu *remoteStorageUtil) uploadOneFile(meta *fileMetadata) error {
+	var encryptMeta *encryptMetadata
+	var dataFile string
+	var err error
+	if meta.encryptionMaterial != nil {
+		if meta.srcStream != nil {
+			var encryptedStream bytes.Buffer
+			srcStream := meta.srcStream
+			if meta.realSrcStream != nil {
+				srcStream = meta.realSrcStream
+			}
+			encryptMeta, err = encryptStream(meta.encryptionMaterial, srcStream, &encryptedStream, 0)
+			if err != nil {
+				return err
+			}
+			meta.realSrcStream = &encryptedStream
+			dataFile = meta.realSrcFileName
+		} else {
+			encryptMeta, dataFile, err = encryptFile(meta.encryptionMaterial, meta.realSrcFileName, 0, meta.tmpDir)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		dataFile = meta.realSrcFileName
+	}
+
+	utilClass := rsu.getNativeCloudType(meta.stageInfo.LocationType)
+	maxConcurrency := int(meta.parallel)
+	var lastErr error
+	maxRetry := defaultMaxRetry
+	for retry := 0; retry < maxRetry; retry++ {
+		if !meta.overwrite {
+			header := utilClass.getFileHeader(meta, meta.dstFileName)
+			if header != nil && meta.resStatus == uploaded {
+				meta.dstFileSize = 0
+				meta.resStatus = skipped
+				return nil
+			}
+		}
+		if meta.overwrite || meta.resStatus == notFoundFile {
+			utilClass.uploadFile(dataFile, meta, encryptMeta, maxConcurrency, meta.options.multiPartThreshold)
+		}
+		if meta.resStatus == uploaded || meta.resStatus == renewToken || meta.resStatus == renewPresignedURL {
+			return nil
+		} else if meta.resStatus == needRetry {
+			if !meta.noSleepingTime {
+				sleepingTime := intMin(int(math.Exp2(float64(retry))), 16)
+				time.Sleep(time.Second * time.Duration(sleepingTime))
+			}
+		} else if meta.resStatus == needRetryWithLowerConcurrency {
+			maxConcurrency = int(meta.parallel) - (retry * int(meta.parallel) / maxRetry)
+			maxConcurrency = intMax(defaultConcurrency, maxConcurrency)
+			meta.lastMaxConcurrency = maxConcurrency
+
+			if !meta.noSleepingTime {
+				sleepingTime := intMin(int(math.Exp2(float64(retry))), 16)
+				time.Sleep(time.Second * time.Duration(sleepingTime))
+			}
+		}
+		lastErr = meta.lastError
+	}
+	if lastErr != nil {
+		return lastErr
+	}
+	return fmt.Errorf("unkown error uploading %v", dataFile)
+}
+
+func (rsu *remoteStorageUtil) uploadOneFileWithRetry(meta *fileMetadata) error {
+	utilClass := rsu.getNativeCloudType(meta.stageInfo.LocationType)
+	retry := true
+	for i := 0; i < 10; i++ {
+		// retry
+		if err := rsu.uploadOneFile(meta); err != nil {
+			return err
+		}
+		if meta.resStatus == uploaded {
+			for j := 0; j < 10; j++ {
+				utilClass.getFileHeader(meta, meta.dstFileName)
+				if meta.resStatus == notFoundFile {
+					time.Sleep(time.Second) // wait 1 second
+					retry = true
+					continue
+				} else {
+					retry = false
+					break
+				}
+			}
+		}
+		if retry {
+			continue
+		} else {
+			break
+		}
+	}
+	if retry {
+		// could not upload file even after retry
+		meta.resStatus = errStatus
```

Comment:
If we got one resStatus = errStatus, we should fail the whole PUT right? But I don't see anywhere you check this 'meta.errStatus' outside, only renewToken and renewnPresignedUrl were checked and handled.

Reply:
As mentioned before, I think we still want to go through with the PUT despite one (or multiple) files fail and return the status of that to the users. The users can then attempt to PUT the files failed files after. If they try to upload the same batch of files, it'll simply return skipped.